include/dgl/immutable_graph.h

```diff
@@ -17,317 +17,444 @@
+  }
+
+  DGLIdIters OutEdgeVec(dgl_id_t vid) const override {
+    const dgl_id_t* indptr_data = static_cast<dgl_id_t*>(indptr_->data);
+    const dgl_id_t* eid_data = static_cast<dgl_id_t*>(edge_ids_->data);
+    const dgl_id_t start = indptr_data[vid];
+    const dgl_id_t end = indptr_data[vid + 1];
+    return DGLIdIters(eid_data + start, eid_data + end);
+  }
+
+  DGLIdIters PredVec(dgl_id_t vid) const override {
+    LOG(FATAL) << "CSR graph does not support efficient PredVec."
+      << " Please use SuccVec on the reverse CSR graph.";
+    return DGLIdIters(nullptr, nullptr);
+  }
+
+  DGLIdIters InEdgeVec(dgl_id_t vid) const override {
+    LOG(FATAL) << "CSR graph does not support efficient InEdgeVec."
+      << " Please use OutEdgeVec on the reverse CSR graph.";
+    return DGLIdIters(nullptr, nullptr);
+  }
+
+  GraphInterface *Reset() override {
+    CSR* gptr = new CSR();
+    *gptr = std::move(*this);
+    return gptr;
+  }
+
+  std::vector<IdArray> GetAdj(bool transpose, const std::string &fmt) const override {
+    CHECK(!transpose && fmt == "csr") << "Not valid adj format request.";
+    return {indptr_, indices_, edge_ids_};
+  }
+
+  /*! \brief Return the reverse of this CSR graph (i.e, a CSC graph) */
+  CSRPtr Transpose() const;
+
+  /*! \brief Convert this CSR to COO */
+  COOPtr ToCOO() const;
+
+  /*!
+   * \return the csr matrix that represents this graph.
+   * \note The csr matrix shares the storage with this graph.
+   *       The data field of the CSR matrix stores the edge ids.
+   */
+  CSRMatrix ToCSRMatrix() const {
+    return CSRMatrix{indptr_, indices_, edge_ids_};
+  }
+
+  // member getters
+
+  IdArray indptr() const { return indptr_; }
+
+  IdArray indices() const { return indices_; }
+
+  IdArray edge_ids() const { return edge_ids_; }
+
+ private:
+  /*! \brief prive default constructor */
+  CSR() {}
+
+  // The CSR arrays.
+  //  - The index is 0-based.
+  //  - The out edges of vertex v is stored from `indices_[indptr_[v]]` to
+  //    `indices_[indptr_[v+1]]` (exclusive).
+  //  - The indices are *not* necessarily sorted.
+  // TODO(minjie): in the future, we should separate CSR and graph. A general CSR
+  //   is not necessarily a graph, but graph operations could be implemented by
+  //   CSR matrix operations. CSR matrix operations would be backed by different
+  //   devices (CPU, CUDA, ...), while graph interface will not be aware of that.
+  IdArray indptr_, indices_, edge_ids_;
 #ifndef _WIN32
-    std::shared_ptr<runtime::SharedMemory> mem;
+  // shared memory handler
+  std::shared_ptr<runtime::SharedMemory> mem_;
 #endif  // _WIN32
-  };
+};
 
-  // Edge list indexed by edge id;
-  struct EdgeList {
-    typedef std::shared_ptr<EdgeList> Ptr;
-    std::vector<dgl_id_t> src_points;
-    std::vector<dgl_id_t> dst_points;
+class COO : public GraphInterface {
+ public:
+  // Create a coo graph that shares the given src and dst
+  COO(int64_t num_vertices, IdArray src, IdArray dst);
```

Comment:
this should also support shared memory. please add a TODO here for me.

Reply:
done